tensorflow/compiler/xla/service/gpu/nvptx_compiler.h

```diff
@@ -35,6 +36,47 @@ void WarnIfBadDriverJITVersion();
 // Returns the directory containing nvvm libdevice files.
 string GetLibdeviceDir(const HloModuleConfig& hlo_module_config);
 
+
+// Persistent compilation cache.
+// This cache stores .ptx and .cubin files to be used by subsequent
+// compilations. The cache is a directory of files. The file name is a hash of
+// the file content. All files are read (disk IO) and stored in memory when the
+// cache is constructed. When an uncached compilation occurs, the result is
+// written (disk IO) to the cache directory immediately. Autotuning is currently
+// non-deterministic, so a few executions might be required to populate the
+// cache.
+// 
+// Deployment:
+// For best performance, keep the cache small (per model) containing only the
+// binaries needed for this execution. In that scenario, after cache creation,
+// there will be no disk IO.
+class PersistentCompilationCache
```

Comment:
The hashing scheme here seems quite tricky and would be difficult to maintain.  is it possible to use a more straightforward: `flat_hash_map<string, std::pair<string, vector<uint8>>>` instead, that maps LLVM IR to PTX and cubin?  The only issue is serializing the (LLVM IR, ptx, cubin) triple right?  I think that would be conceptually simpler than what we have right now.

Reply:
I'd suggest pulling this class out into its own header / cc file, and adding a test.

Reply:
We need more then the LLVM IR, ptx, cubin information. We need to also check the compute capability and the ptx options:

https://github.com/tensorflow/tensorflow/pull/50071/files#diff-ed45b9dd1f983452aae82c67898cccc688fe3ddbc1a00f84125f21e5a97a1dc3R413-R416

I do not understand what would be hard to maintain.
The only difficult part that I see is to make sure that all the information is added to the key. This is needed for both options. Otherwise the code doesn't need maintenance.

Do not forget that this isn't enabled by default and that it target production not research.

Reply:
> I do not understand what would be hard to maintain.

It is difficult to give a crisp answer here since there is no algorithm to decide this (otherwise we software engineers wouldn't have a job :) ), but subjectively here's what I mean:

Right now the data structure does not mirror its intent, so I cannot derive the caching scheme reading the class definition:

```
    static constexpr const int64 kPtxHash = 0xBA55ED50;
    static constexpr const int64 kCubinHash = 0xBA5ADDED;
    bool in_use_;
    string cache_dir_;
    absl::flat_hash_map<int64, string > in_memory_cache_;
```

whereas the caching scheme would be near obvious if we had:

```
  using LlvmIr = std::string;
  using CompiledPtx = std::string;
  using CuBin = std::vector<uint8>;
  absl::flat_hash_map<std::pair<ComputeCapability, LlvmIr>, std::pair<CompiledPtx, CuBin>> cache_;
```

Reply:
This would request a big rewrite of the cache. 
This would complicate the files structure too.
Currently we only cache the value and the hash of the key. Not the full key. So when reading the cache, we can't populate the structure you suggest.

What you propose will move the complexity from understanding the in memory caching scheme to the file caching scheme itself. It doesn't seem a good win.
As it seems to me to just move the complexity around and it is a non-trivial change, I do not like this suggestion.

I added more comments to explain it better. So it will help clarify the current hashing mechanism. Which was your main point.

I also did a small refactoring to make have only 1 place to edit when we change ptxas parameters. This will also help for the maintenance of the code.

What do you think of this new version? Does it explain well the caching structure?